vm/devices/support/fs/lxutil/src/windows/symlink.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use pal::windows::UnicodeString;
+use widestring::U16CString;
+use windows::Wdk::Storage::FileSystem;
+use windows::Win32::System::SystemServices as W32Ss;
+
+/// Get the symlink substitute name and flags from the reparse data.
+fn get_substitute_name(
+    reparse: &FileSystem::REPARSE_DATA_BUFFER,
+) -> lx::Result<(UnicodeString, u32)> {
+    // safety: Caller guarantees that the reparse data buffer is well-formed.
+    let (buffer, offset, length, flags) = unsafe {
+        match reparse.ReparseTag {
+            W32Ss::IO_REPARSE_TAG_SYMLINK => (
+                &reparse.Anonymous.SymbolicLinkReparseBuffer.PathBuffer,
+                reparse
+                    .Anonymous
+                    .SymbolicLinkReparseBuffer
+                    .SubstituteNameOffset,
+                reparse
+                    .Anonymous
+                    .SymbolicLinkReparseBuffer
+                    .SubstituteNameLength,
+                reparse.Anonymous.SymbolicLinkReparseBuffer.Flags,
+            ),
+            W32Ss::IO_REPARSE_TAG_MOUNT_POINT => (
+                &reparse.Anonymous.MountPointReparseBuffer.PathBuffer,
+                reparse
+                    .Anonymous
+                    .MountPointReparseBuffer
+                    .SubstituteNameOffset,
+                reparse
+                    .Anonymous
+                    .MountPointReparseBuffer
+                    .SubstituteNameLength,
+                0,
+            ),
+            _ => return Err(lx::Error::EIO),
+        }
+    };
+
+    // safety: The validity of the reparse buffer is provided by the caller. If the buffer is valid,
+    // the area pointed to by `buffer + offset` is a valid wstring of length `length`, and this operation is safe.
+    let substitute_name = unsafe {
+        UnicodeString::new(std::slice::from_raw_parts(
+            buffer.as_ptr().byte_offset(offset as _),
+            (length as usize) / size_of::<u16>(),
+        ))
+        .map_err(|_| lx::Error::EIO)?
+    };
+
+    Ok((substitute_name, flags))
+}
+
+/// Translates an absolute NT symlink target to an LX path.
+fn translate_absolute_target(
+    substitute_name: &UnicodeString,
+    state: &super::VolumeState,
+) -> lx::Result<UnicodeString> {
+    if state.options.sandbox || state.options.symlink_root.is_empty() {
+        // EPERM is the default return value if no callback is provided
+        return Err(lx::Error::EPERM);
+    }
+
+    // Convert from UTF-16 UNICODE_STRING to String
```

Comment:
Kind of a shame to go through UTF-8, since Windows paths are not always valid UTF-16, and so this translation can fail.

Reply:
Perhaps better to go through `OsString`, and then use `as_encoded_bytes()` as appropriate.